kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/DefaultEventProcessorFactory.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+package org.kie.server.services.jbpm.kafka;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.slf4j.Logger;
+
+import static org.kie.server.services.jbpm.kafka.KafkaServerUtils.getTopicProperty;
+
+class DefaultEventProcessorFactory implements KafkaEventProcessorFactory {
+
+    private static final Logger logger = org.slf4j.LoggerFactory.getLogger(DefaultEventProcessorFactory.class);
+
+    private Map<String, KafkaEventReader> topicReaderMap = new ConcurrentHashMap<>();
+    private Map<ClassLoader, KafkaEventReader> readerInstanceMap = new ConcurrentHashMap<>();
+
+    private Map<String, KafkaEventWriter> topicWriterMap = new ConcurrentHashMap<>();
+    private Map<String, KafkaEventWriter> writerInstanceMap = new ConcurrentHashMap<>();
+
+    private AtomicBoolean defaultWriterInitialized = new AtomicBoolean(false);
+    private KafkaEventWriter defaultWriterInstance;
+
+    public KafkaEventReader getEventReader(String topic, ClassLoader cl) {
+        return topicReaderMap.computeIfAbsent(topic, t -> buildReader(topic, cl));
```

Comment:
you need here to keep in mind that topic is singleton and it could have several class loaders ( different kjars)

Reply:
good catch, thats a bug, I fixed by using a nested map (purpose is to avoid creating potential heavyweight classloader objects)